src/eventreporter/eventbuffer/inmemoryjsoneventbuffer/InMemoryJSONEventBuffer.ts

```
@@ -0,0 +1,438 @@
+      }
+    }
+    this.buffer.unshift(...retryableEvents);
+  }
+
+  private async sendEventImmediately(item: EventData): Promise<void> {
+    const { name, ts, attributes } = item;
+    const event: Ingestion.JSONIngestionEvent = {
+      type: this.type,
+      v: this.v,
+      payloads: [
+        {
+          name,
+          ts,
+          ...attributes,
+        },
+      ],
+    };
+
+    // These are important events which may or may not result into
+    // page re-directs. If a page re-directs, in Safari and Chrome, the network
+    // request shows cancelled but the data reaches the ingestion endpoint.
+    // In Firefox, the request errors out saying 'NS_BINDING_ABORT' state. Hence, add the event
+    // to cncellable events to try with `sendBeacon` lastly.
+    const timestamp = Date.now();
+    if (this.metadata.browserName.toLowerCase() === 'firefox') {
+      this.cancellableEvents.set(timestamp, [event]);
+    }
+
+    let retryCount = 0;
+    let failed = false;
+    let response = null;
+    const body = this.makeRequestBody([event]);
+    try {
+      do {
+        response = await this.send(body);
+        this.cancellableEvents.delete(timestamp);
+        return;
+      } while (
+        response &&
+        !response.ok &&
+        InMemoryJSONEventBuffer.SENDING_FAILURE_CODES.includes(response.status) &&
+        ++retryCount < this.retryCountLimit
+      );
+    } catch (error) {
+      this.logger?.error(`Error in sending immediate event to the ingestion endpoint`);
+      failed = true;
+    }
+    if (retryCount <= this.retryCountLimit) {
+      failed = true;
+    }
+    if (failed) {
+      this.failedIngestionEvents.push(event);
+    }
+  }
+
+  private async send(data: string): Promise<Response> {
+    try {
+      const response = await fetch(this.ingestionURL, {
+        method: 'POST',
+        body: data,
+      });
+      return Promise.resolve(response);
+    } catch (error) {
+      throw Error(error);
```

The `error` is already an Error object (`error instanceof Error === true`). you can just throw an error.

```
throw error;
```

Similarly, I think you can just return a response.

```
return response;
```

Or you may return `Promise.reject(error)` to match the above.

---

Thanks for explanation. Makes sense, I have updated the code now.